virttest/utils_kernel_module.py

```diff
@@ -0,0 +1,117 @@
+                          " loading." % self._module_name)
+        else:
+            cmd = ""
+            if self._was_loaded:
+                cmd += 'rmmod %s; ' % self._module_name
+            cmd += 'modprobe %s %s' % (self._module_name, params)
+            logging.debug("Loading module: %s" % cmd)
+            status, output = process.getstatusoutput(cmd, shell=True,
+                                                     ignore_status=True)
+            if status:
+                raise exceptions.TestError("Couldn't load module %s: %s" % (
+                    self._module_name, output
+                ))
+            else:
+                self._loaded_config = params
+
+    def restore(self):
+        """Restore previous module state"""
+
+        if self._loaded_config:
+            if not self._was_loaded:
+                cmd = 'rmmod %s' % self._module_name
+            else:
+                cmd = 'rmmod %s; modprobe %s %s' % (self._module_name,
+                                                    self._module_name,
+                                                    self._config_backup)
+            logging.debug("Restoring module state: %s" % cmd)
+            status, output = process.getstatusoutput(cmd, shell=True,
+                                                     ignore_status=True)
+            if status:
+                raise exceptions.TestError("Couldn't restore module %s: %s" % (
+                    self._module_name, self._config_backup))
+
+    def _backup_config(self):
+        """
+        Check if self.module_name is loaded and read config
+
+        """
+        config = KernelModuleHandler._load_config(self._module_name)
+        if config:
+            self._config_backup = config
+            self._was_loaded = True
+        else:
+            self._was_loaded = False
+        logging.debug("Backed up %s module state (was_loaded, params)"
+                      "=(%s, %s)" % (self._module_name, self._was_loaded,
+                                     self._config_backup))
+
+    def get_was_loaded(self):
+        """ Read-only property """
+
+        return self._was_loaded
+
+    def get_config_backup(self):
+        """ Read-only property """
+
+        return self._config_backup
+
+    @staticmethod
+    def _load_config(module_name):
+        """
+        Get current module parameters
+
+        :return: Dictionary holding module config {param:value, ...}, None if
+         module not loaded
+        """
+
+        mod_params_path = '/sys/module/%s/parameters/' % module_name
+        if not os.path.exists(mod_params_path):
+            return None
+
+        mod_params = {}
+        params = os.listdir(mod_params_path)
+        for param in params:
+            with open(os.path.join(mod_params_path, param), 'r') as v:
+                mod_params[param] = v.read().strip()
+        return " ".join("%s=%s" % _ for _ in mod_params.items()) if mod_params else ""
```

Comment:
Hmm, this is actually the serialization, but for internal usage the dict of key/values might be actually more useful. How about returning the whole dict and do the serialization in `_backup_config`? In `load_module` the usage of dict should be actually smoother.

Reply:
(anyway again, only a mild suggestion)

Reply:
However, by your review I found a bug in the code from when I remove the dictionary. Thanks.